pkg/controllerhelpers/wait.go

```diff
@@ -31,6 +34,73 @@ type WaitForStateOptions struct {
 	TolerateDelete bool
 }
 
+type AggregatedConditions[Obj runtime.Object] struct {
+	conditions []func(obj Obj) (bool, error)
+	state      []*bool
+}
+
+func NewAggregatedConditions[Obj runtime.Object](condition func(obj Obj) (bool, error), additionalConditions ...func(obj Obj) (bool, error)) *AggregatedConditions[Obj] {
+	conditions := make([]func(obj Obj) (bool, error), 0, 1+len(additionalConditions))
+	conditions = append(conditions, condition)
+	if len(additionalConditions) != 0 {
+		conditions = append(conditions, additionalConditions...)
+	}
+
+	return &AggregatedConditions[Obj]{
+		conditions: conditions,
+		state:      make([]*bool, len(conditions)),
+	}
+}
+
+func (ac AggregatedConditions[Obj]) clearState() {
+	for i := range ac.state {
+		ac.state[i] = nil
+	}
+}
+
+func (ac AggregatedConditions[Obj]) Condition(obj Obj) (bool, error) {
+	ac.clearState()
+
+	allDone := true
+	var err error
+	for i, cond := range ac.conditions {
+		var done bool
+		done, err = cond(obj)
+		ac.state[i] = &done
+		if err != nil {
+			return done, err
```

Comment:
why return early instead of aggregating the errors?

Reply:
It wouldn't change the overall result as you would get an error either way - user shouldn't depend on the value in that case. That is unless you take any side effects into account.

Anyway I'm fine with maintaining the existing behaviour argument.